### DISCUSSION TITLE
Understanding Source Filtering/Loading with Crane

### DISCUSSION THREAD

**Discussion by Gastove** [ipetkov/crane#658] [Q&A] [answered]
Hullo! I've been learning to build Rust projects with crane, and I'm having a really excellent experience. And, there's a thing I really don't understand in the [workspace quickstart flake](https://github.com/ipetkov/crane/blob/master/examples/quick-start-workspace/flake.nix). It demonstrates source loading with `cleanCargoSource`: 

```nix
   src = craneLib.cleanCargoSource ./.;
```

`src` is then `inherit`-ed into `commonArgs`, which in turn are shallow merged into `individualCrateArgs`, which are _then_ shallow merged with each attribute set for building a given workspace. For ease of conversation, I'll copy one here:

```nix
        my-cli = craneLib.buildPackage (individualCrateArgs // {
          pname = "my-cli";
          cargoExtraArgs = "-p my-cli";
          src = fileSetForCrate ./my-cli;
        });
```

...except now we're overriding `src` entirely with the result of `fileSetForCrate`. 

I'm just trying to understand: why? Why not "just" use the `src` from `cleanCargoSource`?

Thanks so much for the help, and the excellent project!

**Comment by ipetkov** [comment, marked as answer]
Hi @Gastove! It is totally fine to use the "entire", cleaned `src` for the workspace, but this has the effect that if you touch any file (kept by the filter) then all derivations which use that same `src` will need to be rebuilt. Imagine you have a workspace of three crates: `foo`, `bar`, and `common`, where `common` is a dependency of both `foo` and `bar`. Touching any of the source files for `foo` should not require rebuilding `bar`, so if you have a slow build it may be more useful to apply the `fileSetForCrate` technique to avoid accidental rebuilds (what this example illustrates is that only the top-level crate's source along with any common code is included for each leaf binary).

If you have a smaller project or don't care too much about the build invalidation, feel free to omit that extra code.

Hope that explains things!
